pkg/credentialprovider/aws/aws_credentials.go

```diff
@@ -63,9 +68,57 @@ func newECRProvider(getterFactory tokenGetterFactory) *ecrProvider {
 	}
 }
 
-// Enabled implements DockerConfigProvider.Enabled. Enabled is true if AWS
-// credentials are found.
+// Enabled implements DockerConfigProvider.Enabled. Enabled is true if we detect
+// an EC2 vm based on checking for the EC2 system UUID, the asset tag (for nitro
+// instances), or instance credentials if the UUID is not present.
+// Ref: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/identify_ec2_instances.html
 func (p *ecrProvider) Enabled() bool {
+	var likelyEC2 bool
+	once.Do(func() {
+		likelyEC2 = tryValidateEC2UUID() || tryValidateEC2Creds()
+	})
+
+	return likelyEC2
+}
+
+func tryValidateEC2UUID() bool {
+	hypervisor_uuid := "/sys/hypervisor/uuid"
+	product_uuid := "/sys/devices/virtual/dmi/id/product_uuid"
+	asset_tag := "/sys/devices/virtual/dmi/id/board_asset_tag"
+
+	if _, err := os.Stat(hypervisor_uuid); err == nil {
+		b, err := ioutil.ReadFile(hypervisor_uuid)
+		if err != nil {
+			klog.Errorf("error checking if this is an EC2 instance: %v", err)
+		} else if string(b[:3]) == "EC2" || string(b[:3]) == "ec2" {
+			klog.V(5).Infof("found 'ec2' in uuid %v from %v, enabling legacy AWS credential provider", string(b), hypervisor_uuid)
+			return true
+		}
+	}
+
+	if _, err := os.Stat(product_uuid); err == nil {
+		b, err := ioutil.ReadFile(product_uuid)
+		if err != nil {
+			klog.Errorf("error checking if this is an EC2 instance: %v", err)
+		} else if string(b[:3]) == "EC2" || string(b[:3]) == "ec2" {
+			klog.V(5).Infof("found 'ec2' in uuid %v from %v, enabling legacy AWS credential provider", string(b), product_uuid)
+			return true
+		}
+	}
+
+	if _, err := os.Stat(asset_tag); err == nil {
+		b, err := ioutil.ReadFile(asset_tag)
+		if err != nil {
+			klog.Errorf("error checking if this is an EC2 instance: %v", err)
+		} else if string(b[:2]) == "i-" {
```

Comment:
I'm not very familiar with this file... does this exist in non-aws environments? is an `i-` prefix really strongly indicative we're in aws?

Reply:
Added a length check as well, so if the file contains only "i-" and is the length of an AWS instance ID, which I think is a pretty good indicator.  Since this only enables the ECR credential provider, I don't think a false positive in an obscure case is that bad--the user would still need to have ECR images in podspecs in order for the `Provide` method to actually be called.